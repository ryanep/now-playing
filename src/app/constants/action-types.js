export const TRACK_CHANGED = 'TRACK_CHANGED';