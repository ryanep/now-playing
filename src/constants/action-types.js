// Tracks
export const TRACK_REQUEST = "TRACK_REQUEST";
export const TRACK_CHANGED = "TRACK_CHANGED";
export const TRACK_UPDATED = "TRACK_UPDATED";
export const TRACK_FAILURE = "TRACK_FAILURE";

// Auth
export const ACCESS_TOKEN_REQUEST = "ACCESS_TOKEN_REQUEST";
export const ACCESS_TOKEN_SUCCESS = "ACCESS_TOKEN_SUCCESS";
export const ACCESS_TOKEN_FAILURE = "ACCESS_TOKEN_FAILURE";
export const REFRESH_TOKEN_REQUEST = "REFRESH_TOKEN_REQUEST";
export const REFRESH_TOKEN_SUCCESS = "REFRESH_TOKEN_SUCCESS";
export const REFRESH_TOKEN_FAILURE = "REFRESH_TOKEN_FAILURE";
