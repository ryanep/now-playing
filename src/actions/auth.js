import * as actionTypes from "../constants/action-types";
import TokenService from "../services/token";
const tokenService = new TokenService();

export function accessTokenRequested() {
  return {
    type: actionTypes.ACCESS_TOKEN_REQUEST
  };
}

export function accessTokenRefreshed({ access_token }) {
  return {
    type: actionTypes.ACCESS_TOKEN_REFRESHED,
    payload: {
      accessToken: access_token
    }
  };
}

export function accessTokenSuccess({ access_token, refresh_token }) {
  return {
    type: actionTypes.ACCESS_TOKEN_SUCCESS,
    payload: {
      accessToken: access_token,
      refreshToken: refresh_token
    }
  };
}

export function accessTokenFailure() {
  return {
    type: actionTypes.ACCESS_TOKEN_FAILURE
  };
}

export function getAccessTokenFromRefresh() {
  return dispatch => {
    dispatch(accessTokenRequested());

    tokenService.getAccessTokenFromRefreshToken().then(({ data, error }) => {
      if (error) {
        return dispatch(accessTokenFailure());
      }

      let { access_token } = data;
      dispatch(accessTokenRefreshed({ access_token }));
    });
  };
}

export function getAccessTokenFromCode(code) {
  return dispatch => {
    dispatch(accessTokenRequested());

    tokenService
      .getAccessTokenFromCode(code)
      .then(({ data, error }) => {
        if (error) {
          return dispatch(accessTokenFailure());
        }

        let { refresh_token, access_token } = data;
        dispatch(accessTokenSuccess({ access_token, refresh_token }));
      })
      .catch(e => console.log(e));
  };
}
