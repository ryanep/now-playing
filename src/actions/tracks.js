import * as actionTypes from "../constants/action-types";
import SpotifyService from "../services/spotify";
const spotifyService = new SpotifyService();

let currentTrackID = 0;

export function getCurrentTrack() {
  return dispatch => {
    dispatch(trackRequested());
    spotifyService
      .getCurrentTrack()
      .then(response => {
        let { item } = response;

        if (currentTrackID !== item.id) {
          dispatch(getCurrentTrackSuccess(response));
          currentTrackID = item.id;
        } else {
          dispatch(trackUpdated(response));
        }
      })
      .catch(err => {
        console.log(err);
        dispatch(getCurrentTrackFailure());
      });
  };
}

export function trackRequested() {
  return {
    type: actionTypes.TRACK_REQUEST
  };
}

export function trackUpdated(track) {
  return {
    type: actionTypes.TRACK_UPDATED,
    track
  };
}

export function getCurrentTrackSuccess(track) {
  return {
    type: actionTypes.TRACK_CHANGED,
    track
  };
}

export function getCurrentTrackFailure() {
  return {
    type: actionTypes.TRACK_REQUEST_FAILURE
  };
}
