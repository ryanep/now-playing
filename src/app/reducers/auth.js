import {
  SPOTIFY_ACCESS_TOKEN,
  SPOTIFY_REFRESH_TOKEN
} from "../constants/storage-keys";

import * as actionTypes from "../constants/action-types";

import { TEST_AUTH } from "../constants/action-types";

const initialState = {
  accessToken: localStorage.getItem(SPOTIFY_ACCESS_TOKEN),
  refreshToken: localStorage.getItem(SPOTIFY_REFRESH_TOKEN)
};

export default function auth(state = initialState, action) {
  switch (action.type) {
    case actionTypes.ACCESS_TOKEN_SUCCESS:
      return {
        ...state,
        accessToken: action.payload.accessToken,
        refreshToken: action.payload.refreshToken
      };
    case actionTypes.ACCESS_TOKEN_REFRESHED:
      return {
        ...state,
        accessToken: action.payload.accessToken
      };
    default:
      return state;
  }
}
