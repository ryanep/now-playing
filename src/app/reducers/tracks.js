import * as types from '../constants/action-types';

const initialState = {
    currentTrack: {},
    previousTracks: []
};

export default function tracks(state = initialState, action) {
    switch (action.type) {
        case types.TRACK_CHANGE:
            return { ...state, currentTrack: action.track };
        default:
            return state;
    }
}