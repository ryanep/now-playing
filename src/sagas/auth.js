import { call, put, takeLatest } from "redux-saga/effects";
import * as authActions from "../actions/auth";
import * as tokenService from "../services/token";
import {
  ACCESS_TOKEN_REQUEST,
  REFRESH_TOKEN_REQUEST
} from "../constants/action-types";

function* getAccessToken({ payload: { code } }) {
  try {
    const { error, access_token, refresh_token } = yield call(
      tokenService.getAccessTokenFromCode,
      code
    );
    if (error) return yield put(authActions.accessTokenFailure(error));
    window.history.pushState({}, document.title, "/");
    yield put(authActions.accessTokenSuccess({ access_token, refresh_token }));
  } catch (error) {
    yield put(authActions.accessTokenFailure(error));
  }
}

function* refreshAccessToken() {
  try {
    const {
      error,
      access_token
    } = tokenService.getAccessTokenFromRefreshToken();
    if (error) return yield put(authActions.refreshTokenFailure(error));
    yield put(authActions.refreshTokenSuccess({ access_token }));
  } catch (error) {
    yield put(authActions.refreshTokenFailure(error));
  }
}

export const authSagas = [
  takeLatest(ACCESS_TOKEN_REQUEST, getAccessToken),
  takeLatest(REFRESH_TOKEN_REQUEST, refreshAccessToken)
];
