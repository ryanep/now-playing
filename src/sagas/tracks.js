import { delay } from 'redux-saga';
import { call, put, takeLatest, select } from 'redux-saga/effects';
import { TRACK_REQUEST } from '../constants/action-types';
import * as trackActions from '../actions/tracks';
import * as authActions from '../actions/auth';
import * as spotifyService from '../services/spotify';

export const currentTrackTitle = state =>
  state.tracks.currentTrack && state.tracks.currentTrack.title;

export function* getCurrentTrack() {
  try {
    const currentTitle = yield select(currentTrackTitle);
    const track = yield call(spotifyService.getCurrentTrack);
    if (currentTitle === track.title) {
      yield put(trackActions.trackUpdated(track));
    } else {
      yield put(trackActions.trackChanged(track));
    }
    yield delay(1000);
    yield call(getCurrentTrack);
  } catch (error) {
    const { status = null } = error;
    if (status === 204) {
      yield put(trackActions.trackNotPlaying());
      yield delay(1000);
      yield call(getCurrentTrack);
    } else if (status === 401) {
      yield put(authActions.accessTokenExpired());
    } else {
      yield put(trackActions.trackFailure(status));
    }
  }
}

export const trackSagas = [takeLatest(TRACK_REQUEST, getCurrentTrack)];
