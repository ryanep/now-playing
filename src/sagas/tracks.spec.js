import { delay } from 'redux-saga';
import { call, put, takeLatest, select } from 'redux-saga/effects';
import * as spotifyService from '../services/spotify';
import * as trackActions from '../actions/tracks';
import * as authActions from '../actions/auth';
import { getCurrentTrack, trackSagas, currentTrackTitle } from './tracks';
import { TRACK_REQUEST } from '../constants/action-types';

describe('sagas/tracks', () => {
  describe('getCurrentTrackTitle selector', () => {
    it('returns null if state.tracks.currentTrack is null', () => {
      const state = { tracks: { currentTrack: null } };
      expect(currentTrackTitle(state)).toEqual(null);
    });

    it('returns the title of the current track', () => {
      const state = { tracks: { currentTrack: { title: '__title__' } } };
      expect(currentTrackTitle(state)).toEqual('__title__');
    });
  });

  const track = {
    title: '__track_new_title__'
  };
  const currentTitle = '__track_title__';
  describe('getCurrentTrack', () => {
    const gen = getCurrentTrack();
    it('selects the currentTrackTitle from the state', () => {
      expect(gen.next('__track_title__').value).toEqual(
        select(currentTrackTitle)
      );
    });

    it('gets the currently playing track from the spotify service', () => {
      expect(gen.next('__track_title__').value).toEqual(
        call(spotifyService.getCurrentTrack)
      );
    });

    it('puts trackChanged action if track title is different from currentTitle', () => {
      expect(gen.next(track).value).toMatchObject(
        put(trackActions.trackChanged(track))
      );
    });

    it('puts trackUpdated action if track title is the smae as currentTitle', () => {
      const gen = getCurrentTrack();
      gen.next();
      gen.next('__track_title__');
      expect(gen.next({ title: '__track_title__' }).value).toEqual(
        put(trackActions.trackUpdated({ title: '__track_title__' }))
      );
    });

    it('calls delay, waiting for 1 second', () => {
      expect(gen.next().value).toMatchSnapshot();
    });

    it('recursively calls getCurrentTrack', () => {
      expect(gen.next().value).toEqual(call(getCurrentTrack));
    });

    describe('error handling', () => {
      describe('api response status 204', () => {
        const gen = getCurrentTrack();
        gen.next();
        it('puts trackNotPlaying action', () => {
          expect(gen.throw({ status: 204 }).value).toEqual(
            put(trackActions.trackNotPlaying())
          );
        });

        it('calls delay, waiting for 1 second', () => {
          expect(gen.next().value).toMatchSnapshot();
        });

        it('recursively calls getCurrentTrack', () => {
          expect(gen.next().value).toEqual(call(getCurrentTrack));
        });
      });

      describe('api response status 401', () => {
        const gen = getCurrentTrack();
        gen.next();
        it('puts accessTokenExpired action', () => {
          expect(gen.throw({ status: 401 }).value).toEqual(
            put(authActions.accessTokenExpired())
          );
        });
      });

      describe('api response with no status', () => {
        const gen = getCurrentTrack();
        gen.next();
        it('sets the status to null if none exists', () => {
          expect(gen.throw({}).value).toEqual(
            put(trackActions.trackFailure(null))
          );
        });
      });

      describe('any other errors', () => {
        const error = { status: 503 };
        const gen = getCurrentTrack();
        gen.next();
        it('puts trackFailure action with status as the argument', () => {
          expect(gen.throw(error).value).toEqual(
            put(trackActions.trackFailure(503))
          );
        });
      });
    });
  });

  describe('trackSagas', () => {
    it('exports an array of sagas', () => {
      expect(trackSagas).toBeInstanceOf(Array);
    });

    it('exposes the expected format', () => {
      expect(trackSagas).toEqual([takeLatest(TRACK_REQUEST, getCurrentTrack)]);
    });
  });
});
