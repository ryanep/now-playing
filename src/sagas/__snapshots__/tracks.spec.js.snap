// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sagas/tracks getCurrentTrack calls delay, waiting for 1 second 1`] = `
Promise {
  "@@redux-saga/CANCEL_PROMISE": [Function],
}
`;

exports[`sagas/tracks getCurrentTrack error handling api response status 204 calls delay, waiting for 1 second 1`] = `
Promise {
  "@@redux-saga/CANCEL_PROMISE": [Function],
}
`;
